text/2978-stack_based_vec.md

```diff
@@ -0,0 +1,281 @@
+
+```rust
+// `array_vec` can store up to 64 elements
+let mut array_vec: ArrayVec<i32, 64> = ArrayVec::new();
+```
+
+Of course, fixed buffers lead to inflexibility because unlike `Vec`, the underlying capacity can not expand at run-time and there will never be more than 64 elements in the above example.
+
+```rust
+// This vector can store up to 0 elements, therefore, nothing at all
+let mut array_vec: ArrayVec<i32, 0> = ArrayVec::new();
+let push_result = array_vec.push(1);
+// Ooppss... Our push operation wasn't successful
+assert!(push_result.is_err());
+```
+
+A good question is: Should I use `core::collections::ArrayVec<T>` or `alloc::collections::Vec<T>`? Well, `Vec` is already good enough for most situations while stack allocation usually shines for small sizes.
+
+* Do you have a known upper bound?
+
+* How much memory are you going to allocate for your program? The default values of `RUST_MIN_STACK` or `ulimit -s` might not be enough.
+
+* Are you using nested `Vec`s? `Vec<ArrayVec<T, N>>` might be better than `Vec<Vec<T>>`.
+
+Each use-case is different and should be pondered individually. In case of doubt, stick with `Vec`.
+
+For a more technical overview, take a look at the following operations:
+
+```rust
+// `array_vec` has a pre-allocated memory of 2048 bits (32 * 64) that can store up
+// to 64 decimals.
+let mut array_vec: ArrayVec<i32, 64> = ArrayVec::new();
+
+// Although reserved, there isn't anything explicitly stored yet
+assert_eq!(array_vec.len(), 0);
+
+// Initializes the first 32 bits with a simple '1' decimal or
+// 00000000 00000000 00000000 00000001 bits
+array_vec.push(1);
+
+// Our vector memory is now split into a 32/2016 pair of initialized and
+// uninitialized memory respectively
+assert_eq!(array_vec.len(), 1);
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`ArrayVec` is a contiguous memory block where elements can be collected, therefore, a collection by definition and even though `core::collections` doesn't exist, it is the most natural module placement.
+
+The API basically mimics most of the current `Vec` surface with some tweaks to manage capacity.
+
+Notably, these tweaked methods are checked (out-of-bound inputs or invalid capacity) versions of some well-known functions like `push` that will return `Result` instead of panicking at run-time. Since the upper capacity bound is known at compile-time and the majority of methods are `#[inline]`, the compiler is likely going to remove most of the conditional bounding checking.
+
+```rust
+// Please, bare in mind that these methods are simply suggestions. Discussions about the
+// API should probably take place elsewhere.
+
+pub struct ArrayVec<T, const N: usize> {
+    data: MaybeUninit<[T; N]>,
```

Comment:
Should this be `[MaybeUninit<T>; N]` or `MaybeUninit<[T; N]>`. I think the former would be better, since it allows simpler implementations of most of the functions, due to it being an array. The only complication is `From::from<[T; N]>`, since generic transmutes don't work yet, and instead have to go through `transmute_copy`

Reply:
In my opinion, `MaybeUninit<[T; N]>` has greater semantic value.

If they generate the same machine code, then `[MaybeUninit<T>; N]` should be preferable because it is more popular than `MaybeUninit<[T; N]>`.

Reply:
I would rather `[MaybeUninit<T>; N]`, since saying `MaybeUninit<[T; N]>` is like saying either *all* of the array or *none* of the array is initialized. It also allows simpler and safer implementations of `push`, `pop`, and other functions, since using pointer offsets isn't needed:

```rust
fn push(&mut self, element: T) -> Result<(), T> {
    if self.len == N {
        Err(T)
    } else {
        self.data[self.len].write(element);
        self.len += 1;
        Ok(())
    }
}

fn pop(&mut self) -> Option<T> {
    if self.len == 0 {
        None
    } else {
        self.len -= 1;
        unsafe { self.data[self.len].assume_init_read() }
    }
}
```

Reply:
@mbartlett21 My biggest concern in this case is about machine code generation or well `[MaybeUninit<T>; N]` performs over `MaybeUninit<[T; N]>`. I just need some time to see if they generate the same assembly, and if they really do, then we can indeed stick with `[MaybeUninit<T>; N]`.

cc https://github.com/rust-lang/rust/pull/81167

Reply:
At the moment, there is a bounds check on the array accesses in the above examples.
If I add this function to the vector, then use it in the functions, it then removes the bound checks:

```rust
#[inline(always)]
fn assume_len_in_bounds(&self) {
    unsafe { core::intrinsics::assume(self.len <= 64) }
}

// The edited methods from above
fn push(&mut self, element: T) -> Result<(), T> {
    self.assume_len_in_bounds();
    if self.len == N {
        Err(T)
    } else {
        self.data[self.len].write(element);
        self.len += 1;
        Ok(())
    }
}

fn pop(&mut self) -> Option<T> {
    if self.len == 0 {
        None
    } else {
        // This has to be before the decrement,
        // since otherwise 64 is a valid value of `len`
        self.assume_len_in_bounds();
        self.len -= 1;
        unsafe { self.data[self.len].assume_init_read() }
    }
}
```

Is there any way that we can say that an invariant of the type is that `self.len <= N`?

Reply:
Maybe `assume_len_in_bounds` could be placed in the constructor?

> Is there any way that we can say that an invariant of the type is that self.len <= N?

Given enough context, the compiler can infer if a branch is "invariable". What Rustc and most people usually do is place something like `assert!(self.len <= N)` at the begging  of the method.

Reply:
I might be wishing for ponies here, but could this somehow be generalized further that the data could either be a `[MaybeUninit<[T]>; N]` or a `Box<[MaybeUninit<[T]>; N]>`? This could make gradually initializing a box more ergonomic. Once `len == N` the fully initialized Box could be extracted without reallocation. Then the user would need zero `unsafe` code.

Reply:
The reason `Box` isn't used is so that there is no heap allocation, and it can also be used in `const` contexts.

If there was an impl for `TryInto<Box<[T; N]>>` for `Box<T>` the boxed case could be handled without either reallocation or `unsafe` code:

```rust
let mut v = Vec::with_capacity(2);
v.push(this);
v.push(that);
let b: Box<[T; 2]> = v.into_boxed_slice().try_into().unwrap();
```

Reply:
That actually works already. Not sure why I didn't think about the vec-box conversions. No ponies needed then.

Reply:
I didn't realize that there was aready an impl of `TryFrom<Box<[T]>>` for `Box<[T; N]>`...